vhost-device-input/src/vhu_input.rs

```diff
@@ -0,0 +1,625 @@
+        let mut name: [u8; VIRTIO_INPUT_CFG_SIZE] = [0; VIRTIO_INPUT_CFG_SIZE];
+
+        // SAFETY: Safe as the file is a valid event device, the kernel will only
+        // update the correct amount of memory in func.
+        match unsafe { eviocgname(self.ev_dev.get_raw_fd(), name.as_mut_slice()) } {
+            Ok(len) if len as usize > name.len() => {
+                return Err(VuInputError::UnexpectedInputDeviceError);
+            }
+            Ok(len) if len <= 1 => {
+                return Err(VuInputError::UnexpectedInputDeviceError);
+            }
+            Err(_) => {
+                return Err(VuInputError::UnexpectedInputDeviceError);
+            }
+            _ => (),
+        }
+
+        let size = String::from_utf8(name.to_vec()).unwrap().len();
+
+        Ok(VuInputConfig {
+            select: self.select,
+            subsel: 0,
+            size: size as u8,
+            reserved: [0; 5],
+            val: name,
+        })
+    }
+
+    pub fn read_id_config(&self) -> Result<VuInputConfig> {
+        let input_id = self.ev_dev.input_id();
+
+        let mut dev_id = [
+            input_id.bus_type().0.as_slice(),
+            input_id.vendor().as_slice(),
+            input_id.product().as_slice(),
+            input_id.version().as_slice(),
+        ]
+        .concat();
+
+        dev_id.resize(VIRTIO_INPUT_CFG_SIZE, 0);
+
+        Ok(VuInputConfig {
+            select: VIRTIO_INPUT_CFG_ID_DEVIDS,
+            subsel: 0,
+            size: VIRTIO_INPUT_CFG_SIZE as u8,
+            reserved: [0; 5],
+            val: dev_id.try_into().unwrap(),
+        })
+    }
+}
+
+/// VhostUserBackend trait methods
+impl<T: 'static + InputDevice + Sync + Send> VhostUserBackendMut<VringRwLock, ()>
+    for VuInputBackend<T>
+{
+    fn num_queues(&self) -> usize {
+        NUM_QUEUES
+    }
+
+    fn max_queue_size(&self) -> usize {
+        QUEUE_SIZE
+    }
+
+    fn features(&self) -> u64 {
+        1 << VIRTIO_F_VERSION_1
+            | 1 << VIRTIO_RING_F_INDIRECT_DESC
+            | 1 << VIRTIO_RING_F_EVENT_IDX
+            | VhostUserVirtioFeatures::PROTOCOL_FEATURES.bits()
+    }
+
+    fn protocol_features(&self) -> VhostUserProtocolFeatures {
+        VhostUserProtocolFeatures::MQ | VhostUserProtocolFeatures::CONFIG
+    }
+
+    fn get_config(&self, _offset: u32, size: u32) -> Vec<u8> {
+        let cfg = match self.select {
+            VIRTIO_INPUT_CFG_ID_NAME => self.read_name_config(),
+            VIRTIO_INPUT_CFG_ID_DEVIDS => self.read_id_config(),
+            VIRTIO_INPUT_CFG_EV_BITS => self.read_event_config(),
+            _ => {
+                // Return zeros for unsupported config types
+                return vec![0; size as usize];
+            }
+        };
+
+        match cfg {
+            Ok(val) => val.to_bytes(),
+            _ => {
+                vec![0; size as usize]
+            }
+        }
+    }
+
+    fn set_config(&mut self, _offset: u32, buf: &[u8]) -> io::Result<()> {
```

Comment:
offset needs to be used. A driver may not do issue writes for the whole area at once.

Reply:
For input device, this is not true for using "offset" when invorking set_config().  See the C version's code: https://gitlab.com/qemu-project/qemu/-/blob/master/contrib/vhost-user-input/main.c?ref_type=heads#L231

Reply:
If we look at spec https://docs.oasis-open.org/virtio/virtio/v1.2/cs01/virtio-v1.2-cs01.pdf, section "5.8.5.1 Driver Requirements: Device Initialization", it doesn't mention it uses `offset` parameter.

```
A driver MUST set both select and subsel when querying device configuration, in any order.
A driver MUST NOT write to configuration fields other than select and subsel.
A driver SHOULD check the size field before accessing the configuration information.
```

Reply:
@Leo-Yan putting this in the function as a comment along with the spec version and URL would be a good idea.